Secure socket path

millerdev opened the pull request:
Moved the socket path to a location that can be locked down so only the xremap process and the owning user can access it.

The new socket location is most easily managed with a socket routing service. [`xremap-socket`](https://github.com/millerdev/xremap-socket), which monitors user sessions and routes active window requests to the active user session, was written for that very purpose.

millerdev committed:
Move socket path to a more secure location

Default: /run/xremap/{UID}/gnome.sock
Can be overridden with environment variable: XREMAP_GNOME_SOCKET

The default is intended to be a directory that is accessible only by the
user running xremap as well as the user identified by UID.

millerdev committed:
Remove settings - not needed

millerdev committed:
Remove obsolete _isLocked property

millerdev committed:
Add socket path to relevant error logs

millerdev committed:
Add configuration instructions to README

k0kubun commented:
Can you leave https://github.com/xremap/xremap-gnome/pull/15 separated from this PR please? I'd like to review/merge them independently.

millerdev commented:
Re-opened #15. Once that's merged I will rebase this one. Otherwise there would be conflicts.

k0kubun commented:
Merged that one. Please rebase this PR :pray:

millerdev committed:
Version 13

millerdev commented:
Rebased.

millerdev commented:
@k0kubun any reservations on getting this merged?

hpccc53 commented:
If I may offer my opinion.

My reservations is that xremap-socket becomes kinda official, when linked to from documentation.

Also, an extra program makes the solution more complex and inconvenient for users. It's seems to me the extra program is the only reason for two services and rerouting of sockets.

So I suggest xremap-socket is incrementally introduced into main xremap, so it's easier to maintain, and easy to install before recommended to users.

I believe the only thing, that needs to interface with the detect-seated-user functionality is the GnomeClient, which can do it via a channel to another thread, which then can do async as it wants.

Regarding the GNOME extension. If support is added for `/run/xremap/${uid}/gnome.sock` then I think it should be stated as experimental and opt-in. The default should remain `/run/xremap/gnome.sock`.

k0kubun commented:
> xremap-socket becomes kinda official, when linked to from documentation

This is what I thought too. If we merge this, I want xremap-socket to be transferred to xremap org and maintained together with the other extension.

But first of all, I'm not sure if I understand why they need to be separate extensions. Can they be implemented as a single extension? It's just tedious for users to install two different extensions. If necessary, we can maybe make the socket capability an optional feature?

millerdev commented:
> Regarding the GNOME extension. If support is added for `/run/xremap/${uid}/gnome.sock` then I think it should be stated as experimental and opt-in. The default should remain `/run/xremap/gnome.sock`.

I don't think `/run/xremap/gnome.sock` is a good default because it is insecure: it's vulnerable to race conditions and denial of service attacks. If permissions are not configured properly, it allows any user to snoop on the GNOME session of the current socket owner.

> I want xremap-socket to be transferred to xremap org and maintained together with the other extension.

That's fine, I'm happy to transfer it. What's the process for doing that?

> Can they be implemented as a single extension? It's just tedious for users to install two different extensions.

This is a good question, and I agree, having to maintain a separate socket manager process adds unnecessary complexity. However, xremap-socket cannot be merged into the GNOME extension because the GNOME extension runs as the current user, and xremap-socket must run as a user that has permission to read all user sockets. I think it would make most sense to merge xremap-socket into main xremap to reduce complexity.

hpccc53 commented:
> I don't think /run/xremap/gnome.sock is a good default because it is insecure: it's vulnerable to race conditions and denial of service attacks. If permissions are not configured properly, it allows any user to snoop on the GNOME session of the current socket owner.

I agree with you on that. I only suggested it to be able to move forward with this PR, to allow early adopters of `/run/xremap/${uid}/gnome.sock` while `xremap-socket` was made convenient. As you state, it is an improvement.

What about first checking if /run/xremap/${uid} exists and then behave the new way. Otherwise fallback to the current. 

I imagine this is future-proof, we don't have to make breaking changes later. And when users upgrade the xremap-binary the gnome extension automatically picks that up.

PS: `/run/xremap/${uid}/xremap.sock` would be better, because it can be used for other desktops as well.

millerdev committed:
Update default socket path

Default: /run/xremap/{UID}/xremap.sock (if parent directory exists)
Legacy default: /run/xremap/gnome.sock (fallback if /run/xremap/{UID} does not exist)
Can be overridden with environment variable: XREMAP_GNOME_SOCKET

millerdev committed:
Remove xremap-socket reference from docs

millerdev commented:
Updated:

Default: `/run/xremap/{UID}/xremap.sock` (if parent directory exists)
Legacy default: `/run/xremap/gnome.sock` (fallback if `/run/xremap/{UID}` does not exist)
Can be overridden with environment variable: `XREMAP_GNOME_SOCKET`

hpccc53 commented:
I think that works. I can't merge here, so I'll defer to @k0kubun to make the decision.

millerdev commented:
Once merged, this will work with https://github.com/xremap/xremap/pull/827

k0kubun commented:
works for me too :+1:

> That's fine, I'm happy to transfer it. What's the process for doing that?

You will need to join this organization to have the transfer permission. I have invited you. Could you accept it and transfer the repository when you get a chance?

> I think that works. I can't merge here, so I'll defer to @k0kubun to make the decision.

With the amount of contributions you have given to xremap, I want you to be able to merge changes here as well. For simplicity, I just gave permission on xremap-gnome to the xremap team (which includes you).

I will merge and release the patch once the repository is transferred.

millerdev commented:
> You will need to join this organization to have the transfer permission. I have invited you. Could you accept it and transfer the repository when you get a chance?

Got it, thank you.

I propose that it makes more sense to _not transfer_ at this point it since [the socket client has been added to xremap](https://github.com/xremap/xremap/pull/827). I think having [xremap-socket](https://github.com/millerdev/xremap-socket) in the xremap org would probably just create confusion. I have added a note to the xremap-socket README that it is obsolete. Let me know if you still think it's valuable and want me to do the transfer.

millerdev commented:
> I will merge and release the patch once the repository is transferred.

Sorry, I was reading too fast and missed this note. I hope it's not a problem that I went ahead and merged this PR.

millerdev commented:
I am unable to do the last step in the release process: Open https://extensions.gnome.org/upload/ and upload it.

@k0kubun I assume you have to do it since you own the GNOME extension.

k0kubun commented:
Sure, no problem. I've just uploaded the latest master https://github.com/xremap/xremap-gnome/commit/87571273f2746176988e8fb671f5576d3407054c as Version 13.

Since you decided not to use the transfer permission, I demoted you from an owner to a member. However, now that you're in the org, I decided to welcome you as an maintainer of xremap-gnome as well as xremap. Feel free to say no if you're not interested.

millerdev commented:
I appreciate it, thank you.
